File: src/copaw/app/routers/fixed_sops.py
# -*- coding: utf-8 -*-
from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query, Request

from ...sop_kernel import (
    FixedSopBindingCreateRequest,
    FixedSopBindingDetail,
    FixedSopDoctorReport,
    FixedSopRunDetail,
    FixedSopRunRequest,
    FixedSopRunResponse,
    FixedSopService,
    FixedSopTemplateListResponse,
)
from ...state import FixedSopTemplateRecord

router = APIRouter(prefix="/fixed-sops", tags=["fixed-sops"])


def _get_fixed_sop_service(request: Request) -> FixedSopService:
    service = getattr(request.app.state, "fixed_sop_service", None)
    if isinstance(service, FixedSopService):
        return service
    raise HTTPException(503, detail="Fixed SOP service is not available")


@router.get("/templates", response_model=FixedSopTemplateListResponse)
async def list_fixed_sop_templates(
    request: Request,
    status: str | None = Query(default="active"),
) -> FixedSopTemplateListResponse:
    return _get_fixed_sop_service(request).list_template_catalog(status=status)


@router.get("/templates/{template_id}", response_model=FixedSopTemplateRecord)
async def get_fixed_sop_template(
    request: Request,
    template_id: str,
) -> FixedSopTemplateRecord:
    template = _get_fixed_sop_service(request).get_template(template_id)
    if template is None:
        raise HTTPException(404, detail=f"Fixed SOP template '{template_id}' not found")
    return template


@router.get("/bindings", response_model=list[FixedSopBindingDetail])
async def list_fixed_sop_bindings(
    request: Request,
    template_id: str | None = Query(default=None),
    status: str | None = Query(default=None),
    industry_instance_id: str | None = Query(default=None),
    owner_agent_id: str | None = Query(default=None),
    limit: int | None = Query(default=50),
) -> list[FixedSopBindingDetail]:
    return _get_fixed_sop_service(request).list_binding_details(
        template_id=template_id,
        status=status,
        industry_instance_id=industry_instance_id,
        owner_agent_id=owner_agent_id,
        limit=limit,
    )


@router.post("/bindings", response_model=FixedSopBindingDetail, status_code=201)
async def create_fixed_sop_binding(
    request: Request,
    payload: FixedSopBindingCreateRequest,
) -> FixedSopBindingDetail:
    service = _get_fixed_sop_service(request)
    try:
        return service.create_binding(payload)
    except KeyError as exc:
        raise HTTPException(404, detail=str(exc).strip("'")) from exc


@router.get("/bindings/{binding_id}", response_model=FixedSopBindingDetail)
async def get_fixed_sop_binding(
    request: Request,
    binding_id: str,
) -> FixedSopBindingDetail:
    service = _get_fixed_sop_service(request)
    try:
        return service.get_binding(binding_id)
    except KeyError as exc:
        raise HTTPException(404, detail=str(exc).strip("'")) from exc


@router.put("/bindings/{binding_id}", response_model=FixedSopBindingDetail)
async def update_fixed_sop_binding(
    request: Request,
    binding_id: str,
    payload: FixedSopBindingCreateRequest,
) -> FixedSopBindingDetail:
    service = _get_fixed_sop_service(request)
    try:
        return service.update_binding(binding_id, payload)
    except KeyError as exc:
        raise HTTPException(404, detail=str(exc).strip("'")) from exc


@router.post("/bindings/{binding_id}/doctor", response_model=FixedSopDoctorReport)
async def run_fixed_sop_doctor(
    request: Request,
    binding_id: str,
) -> FixedSopDoctorReport:
    service = _get_fixed_sop_service(request)
    try:
        return service.run_doctor(binding_id)
    except KeyError as exc:
        raise HTTPException(404, detail=str(exc).strip("'")) from exc


@router.post("/bindings/{binding_id}/run", response_model=FixedSopRunResponse)
async def run_fixed_sop_binding(
    request: Request,
    binding_id: str,
    payload: FixedSopRunRequest,
) -> FixedSopRunResponse:
    service = _get_fixed_sop_service(request)
    try:
        return await service.run_binding(binding_id, payload)
    except KeyError as exc:
        raise HTTPException(404, detail=str(exc).strip("'")) from exc


@router.get("/runs/{run_id}", response_model=FixedSopRunDetail)
async def get_fixed_sop_run(
    request: Request,
    run_id: str,
) -> FixedSopRunDetail:
    service = _get_fixed_sop_service(request)
    try:
        return service.get_run(run_id)
    except KeyError as exc:
        raise HTTPException(404, detail=str(exc).strip("'")) from exc

File: console/src/api/modules/fixedSops.ts
import { request } from "../request";

export interface FixedSopTemplateRecord {
  template_id: string;
  name: string;
  summary: string;
  description: string;
  status: string;
  version: string;
  source_kind: string;
  source_ref?: string | null;
  owner_role_id?: string | null;
  suggested_role_ids: string[];
  industry_tags: string[];
  capability_tags: string[];
  risk_baseline: string;
  input_schema: Record<string, unknown>;
  output_schema: Record<string, unknown>;
  writeback_contract: Record<string, unknown>;
  node_graph: Array<Record<string, unknown>>;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface FixedSopTemplateSummary {
  template: FixedSopTemplateRecord;
  binding_count: number;
  routes: Record<string, string>;
}

export interface FixedSopTemplateListResponse {
  items: FixedSopTemplateSummary[];
  total: number;
}

export interface FixedSopBindingRecord {
  binding_id: string;
  template_id: string;
  binding_name: string;
  status: string;
  owner_scope?: string | null;
  owner_agent_id?: string | null;
  industry_instance_id?: string | null;
  workflow_template_id?: string | null;
  trigger_mode: string;
  trigger_ref?: string | null;
  input_mapping: Record<string, unknown>;
  output_mapping: Record<string, unknown>;
  timeout_policy: Record<string, unknown>;
  retry_policy: Record<string, unknown>;
  risk_baseline: string;
  last_run_id?: string | null;
  last_verified_at?: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface FixedSopBindingDetail {
  binding: FixedSopBindingRecord;
  template: FixedSopTemplateRecord;
  routes: Record<string, string>;
}

export interface FixedSopBindingCreatePayload {
  template_id: string;
  binding_name?: string;
  status?: string;
  owner_scope?: string;
  owner_agent_id?: string;
  industry_instance_id?: string;
  workflow_template_id?: string;
  trigger_mode?: string;
  trigger_ref?: string;
  input_mapping?: Record<string, unknown>;
  output_mapping?: Record<string, unknown>;
  timeout_policy?: Record<string, unknown>;
  retry_policy?: Record<string, unknown>;
  risk_baseline?: string;
  metadata?: Record<string, unknown>;
}

export interface FixedSopDoctorCheck {
  key: string;
  label: string;
  status: "pass" | "warn" | "fail" | "info";
  message: string;
}

export interface FixedSopDoctorReport {
  binding_id: string;
  template_id: string;
  status: "ready" | "degraded" | "blocked";
  summary: string;
  checks: FixedSopDoctorCheck[];
  environment_id?: string | null;
  session_mount_id?: string | null;
  host_requirement: Record<string, unknown>;
  host_preflight: Record<string, unknown>;
  routes: Record<string, string>;
}

export interface FixedSopRunRequest {
  input_payload?: Record<string, unknown>;
  workflow_run_id?: string;
  owner_agent_id?: string;
  owner_scope?: string;
  environment_id?: string;
  session_mount_id?: string;
  dry_run?: boolean;
  metadata?: Record<string, unknown>;
}

export interface FixedSopRunResponse {
  binding_id: string;
  status: "success" | "error";
  summary: string;
  workflow_run_id?: string | null;
  evidence_id?: string | null;
  routes: Record<string, string>;
}

export interface WorkflowRunRecord {
  run_id: string;
  template_id: string;
  title: string;
  summary: string;
  status: string;
  owner_scope?: string | null;
  owner_agent_id?: string | null;
  industry_instance_id?: string | null;
  parameter_payload: Record<string, unknown>;
  preview_payload: Record<string, unknown>;
  goal_ids: string[];
  schedule_ids: string[];
  task_ids: string[];
  decision_ids: string[];
  evidence_ids: string[];
  metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface FixedSopRunDetail {
  run: WorkflowRunRecord;
  binding?: FixedSopBindingRecord | null;
  template?: FixedSopTemplateRecord | null;
  environment_id?: string | null;
  session_mount_id?: string | null;
  host_requirement: Record<string, unknown>;
  host_preflight: Record<string, unknown>;
}

export const fixedSopsApi = {
  listFixedSopTemplates: (params?: { status?: string }) => {
    const search = new URLSearchParams();
    if (params?.status) {
      search.set("status", params.status);
    }
    const suffix = search.toString();
    return request<FixedSopTemplateListResponse>(
      `/fixed-sops/templates${suffix ? `?${suffix}` : ""}`,
    );
  },

  getFixedSopTemplate: (templateId: string) =>
    request<FixedSopTemplateRecord>(
      `/fixed-sops/templates/${encodeURIComponent(templateId)}`,
    ),

  listFixedSopBindings: (params?: {
    template_id?: string;
    status?: string;
    industry_instance_id?: string;
    owner_agent_id?: string;
    limit?: number;
  }) => {
    const search = new URLSearchParams();
    if (params?.template_id) {
      search.set("template_id", params.template_id);
    }
    if (params?.status) {
      search.set("status", params.status);
    }
    if (params?.industry_instance_id) {
      search.set("industry_instance_id", params.industry_instance_id);
    }
    if (params?.owner_agent_id) {
      search.set("owner_agent_id", params.owner_agent_id);
    }
    if (typeof params?.limit === "number") {
      search.set("limit", String(params.limit));
    }
    const suffix = search.toString();
    return request<FixedSopBindingDetail[]>(
      `/fixed-sops/bindings${suffix ? `?${suffix}` : ""}`,
    );
  },

  getFixedSopBinding: (bindingId: string) =>
    request<FixedSopBindingDetail>(
      `/fixed-sops/bindings/${encodeURIComponent(bindingId)}`,
    ),

  createFixedSopBinding: (payload: FixedSopBindingCreatePayload) =>
    request<FixedSopBindingDetail>("/fixed-sops/bindings", {
      method: "POST",
      body: JSON.stringify(payload),
    }),

  updateFixedSopBinding: (
    bindingId: string,
    payload: FixedSopBindingCreatePayload,
  ) =>
    request<FixedSopBindingDetail>(
      `/fixed-sops/bindings/${encodeURIComponent(bindingId)}`,
      {
        method: "PUT",
        body: JSON.stringify(payload),
      },
    ),

  runFixedSopDoctor: (bindingId: string) =>
    request<FixedSopDoctorReport>(
      `/fixed-sops/bindings/${encodeURIComponent(bindingId)}/doctor`,
      {
        method: "POST",
      },
    ),

  runFixedSopBinding: (bindingId: string, payload: FixedSopRunRequest) =>
    request<FixedSopRunResponse>(
      `/fixed-sops/bindings/${encodeURIComponent(bindingId)}/run`,
      {
        method: "POST",
        body: JSON.stringify(payload),
      },
    ),

  getFixedSopRun: (runId: string) =>
    request<FixedSopRunDetail>(`/fixed-sops/runs/${encodeURIComponent(runId)}`),
};
